{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "seasonal-decrease",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt\n",
    "import math\n",
    "import re\n",
    "import warnings\n",
    "import textwrap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "missing-bobby",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "lyric-czech",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read gene information\n",
    "genes = r\"path/genes.csv\"\n",
    "genes = pd.read_csv(genes)\n",
    "\n",
    "# Read masks\n",
    "msk = r\"path/files/mask_def.tsv\"\n",
    "msk = pd.read_csv(msk, sep=\"\\t\", header=None)\n",
    "\n",
    "# Read and prepare list of binary phenotypes\n",
    "phen_b = r\"path/pheno_binary.txt\"\n",
    "pheno_b = pd.read_csv(phen_b, header=None)\n",
    "pheno_b[0] = pheno_b[0].str.rsplit('.',n=1).str.get(0)\n",
    "\n",
    "# Read and prepare list of quantitative phenotypes\n",
    "phen_q = r\"path/pheno_quant.txt\"\n",
    "pheno_q = pd.read_csv(phen_q, header=None)\n",
    "pheno_q[0]=pheno_q[0].str.rsplit('.',n=1).str.get(0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "clear-friday",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pos(s):\n",
    "    '''\n",
    "    function for getting position from snp string\n",
    "    '''\n",
    "    if isinstance(s, str) == False:\n",
    "        return None\n",
    "    if len(s.split(':'))<2:\n",
    "        return None\n",
    "    return int(s.split(':')[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "extreme-celebrity",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_full_inf(pheno, df, phen_type:str, j):\n",
    "    \"\"\"\n",
    "    Function for adding some information (midpoint pos, phenotype, mask name) to df.\n",
    "    -----\n",
    "    pheno: str\n",
    "        the phenotype file to read\n",
    "    df: pandas.DataFrame\n",
    "        df to add results to\n",
    "    phen_type:str\n",
    "        binary or quantitative (for folder address and to fill column)\n",
    "    j: int\n",
    "        chromosome\n",
    "    ------------\n",
    "    return pd.DataFrame\n",
    "    \"\"\"\n",
    " \n",
    "    for i in range(len(pheno)):\n",
    "        # Get phenotype value        \n",
    "        phen = pheno[0].values[i]\n",
    "\n",
    "        # Read regenie output file for chromosome by chromosome and phenotype \n",
    "        rege = r\"path/\"+phen_type+\"/\"+phen+\"/step2_\"+str(j)+\"_\"+phen+\".regenie\"\n",
    "        reg = pd.read_csv(rege, sep=\" \", skiprows=1)\n",
    "        \n",
    "        # Add phenotype column to the regenie output file      \n",
    "        reg['pheno'] = phen\n",
    "        reg['pheno_type'] = phen_type\n",
    "\n",
    "        # Remove all summary rows and reset index\n",
    "        reg = reg.dropna(subset=['ALLELE0'])\n",
    "        reg = reg.reset_index()\n",
    "\n",
    "        # Read snplist file\n",
    "        snpl = r\"path/\"+phen_type+\"/\"+phen+\"/step2_\"+str(j)+\"_masks.snplist\"\n",
    "        snps = pd.read_csv(snpl, sep=\"\\t\", header=None)\n",
    "\n",
    "        # Add info from file to regenie df (number of vars and start and end pos for each row)\n",
    "        snps[1] = snps[1].str.split(',').apply(lambda x: [k for k in x])\n",
    "        reg['enp_s'] = snps[1].apply(lambda x: x[0]).apply(get_pos)\n",
    "        reg['snp_e'] = snps[1].apply(lambda x: x[-1]).apply(get_pos)\n",
    "        reg['pos'] = reg['enp_s'] + (reg['snp_e'] - reg['enp_s'])//2\n",
    "        \n",
    "        # Add mask name        \n",
    "        reg['mask']=''\n",
    "        mask = reg['ALLELE1'].str.rsplit('.',n=2).str.get(0)\n",
    "        for i in range(len(reg)):\n",
    "            if mask.iloc[i]=='Mask1':\n",
    "                reg['mask'].iloc[i] = msk[1].iloc[0]\n",
    "            elif mask.iloc[i]=='Mask2':\n",
    "                reg['mask'].iloc[i] = msk[1].iloc[1]\n",
    "            elif mask.iloc[i]=='Mask3':\n",
    "                reg['mask'].iloc[i] = msk[1].iloc[2]\n",
    "            elif mask.iloc[i]=='Mask4':\n",
    "                reg['mask'].iloc[i] = msk[1].iloc[3]\n",
    "            elif mask.iloc[i]=='Mask5':\n",
    "                reg['mask'].iloc[i] = msk[1].iloc[4]\n",
    "        \n",
    "        \n",
    "        # Add significant results (log10 p-val > cut-off) to the sig df    \n",
    "        df = pd.concat([df, reg])\n",
    "        \n",
    "        \n",
    "    return df\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "handy-germany",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nearest_genes(chrom, pos, genes, distance_le):\n",
    "    \"\"\"\n",
    "    A function for finding the nearest gene to a given\n",
    "    chromosome position.\n",
    "    \n",
    "    ---------------------\n",
    "    chrom int or str:\n",
    "        The chromosome number\n",
    "    pos int:\n",
    "        The position on the chromosome\n",
    "    gene pandas.DataFrame:\n",
    "        Reference dataset showing type and position of genes.\n",
    "    distance_le int:\n",
    "        Maximum the distance to search from our position.\n",
    "    \n",
    "    ---------------------\n",
    "    filt.iloc[0] pandas.core.series.Series: \n",
    "        A pandas series containing the information of the \n",
    "        nearest gene.\n",
    "    pos_st str:\n",
    "        The start position for search.\n",
    "    pos_en str:\n",
    "        The end position for search.\n",
    "        \n",
    "    \"\"\"\n",
    "    pos_st = pos - distance_le\n",
    "    pos_en = pos + distance_le\n",
    "\n",
    "    filt = genes[genes['genetype']==\"protein_coding\"]\n",
    "    filt=filt.dropna(subset=['hgncid'])\n",
    "    filt = filt[filt['chrom']==str(chrom)]\n",
    "    filt = filt[filt['pos_end']>=pos_st]\n",
    "    filt = filt[filt['pos_start']<=pos_en]\n",
    "\n",
    "    if(len(filt)==0):\n",
    "        return None, str(pos_st), str(pos_en)\n",
    "    \n",
    "    filt['dist2gene'] = 0.0\n",
    "    for j in range(len(filt)):\n",
    "        if filt['pos_start'].iloc[j] >= pos:  \n",
    "            filt['dist2gene'].iloc[j] = float(filt['pos_start'].iloc[j] - pos)\n",
    "        elif filt['pos_end'].iloc[j] >= pos and filt['pos_start'].iloc[j] <= pos:\n",
    "            filt['dist2gene'].iloc[j] = 0\n",
    "        elif filt['pos_end'].iloc[j] <= pos:\n",
    "            filt['dist2gene'].iloc[j] = float(pos - filt['pos_end'].iloc[j])    \n",
    "    \n",
    "    filt = filt.sort_values(by=['dist2gene'])\n",
    "    return filt.iloc[0], str(pos_st), str(pos_en)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "liquid-professor",
   "metadata": {},
   "outputs": [],
   "source": [
    "def snp_inf(pheno, reg, phen_type:str):\n",
    "    '''\n",
    "    Fill more snp info.\n",
    "    '''\n",
    "    for j in [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,'X']:\n",
    "        for i in range(len(pheno)):\n",
    "            # Get phenotype value        \n",
    "            phen = pheno[0].values[i]\n",
    "\n",
    "            # Read snplist file\n",
    "            snpl = r\"/path/\"+phen_type+\"/\"+phen+\"/step2_\"+str(j)+\"_masks.snplist\"\n",
    "            snps = pd.read_csv(snpl, sep=\"\\t\", header=None)\n",
    "\n",
    "            # Add info from file to regenie df (number of vars and start and end pos for each row)\n",
    "            snps[1] = snps[1].str.split(',').apply(lambda x: [k for k in x])\n",
    "            snps['num_var'] = snps[1].apply(len)\n",
    "            snps['snp_start_pos'] = snps[1].apply(lambda x: x[0])\n",
    "            snps['snp_end_pos'] = snps[1].apply(lambda x: x[-1])\n",
    "            snps = snps.rename(columns={0:'ID'})\n",
    "            snps = snps.drop([1],axis=1)\n",
    "            snps = snps.set_index('ID')\n",
    "            reg['num_var']=reg['ID'].map(snps['num_var']).fillna(reg['num_var'])\n",
    "            reg['snp_start_pos']=reg['ID'].map(snps['snp_start_pos']).fillna(reg['snp_start_pos'])\n",
    "            reg['snp_end_pos']=reg['ID'].map(snps['snp_end_pos']).fillna(reg['snp_end_pos'])\n",
    "            snps = None\n",
    "        \n",
    "    return reg\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "exterior-potato",
   "metadata": {},
   "outputs": [],
   "source": [
    "def other_inf(sig):\n",
    "    '''\n",
    "    Function for getting some other info from existing columns and logs.\n",
    "    '''\n",
    "    # This cell is for adding column to sig to use for plots\n",
    "    sig['CHROM']=sig['CHROM'].replace(23, 'X')\n",
    "    sig['CHROM']=sig['CHROM'].replace(23.0, 'X')\n",
    "\n",
    "    # Init columns\n",
    "\n",
    "    sig['Al1']=''\n",
    "    sig['OR']=0\n",
    "    sig['upper']=0\n",
    "    sig['lower']=0\n",
    "    sig['MAF']=0\n",
    "    sig['ID2']=''\n",
    "    sig['cases']=0\n",
    "    sig['controls']=0\n",
    "\n",
    "    # Init column based values (sub strings)\n",
    "    ID = sig['ID'].str.rsplit('.',n=3).str.get(0)\n",
    "    maf = sig['ALLELE1'].str.split('.',n=1).str.get(1)\n",
    "    for i in range(len(sig)):\n",
    "        # Fill other initialized columns   \n",
    "        sig['OR'].iloc[i] = math.exp(sig['BETA'].iloc[i])\n",
    "        sig['upper'].iloc[i] = math.exp(sig['BETA'].iloc[i]+(1.96*sig['SE'].iloc[i]))\n",
    "        sig['lower'].iloc[i] = math.exp(sig['BETA'].iloc[i]-(1.96*sig['SE'].iloc[i]))\n",
    "        sig['MAF'].iloc[i] = maf.iloc[i]\n",
    "        sig['Al1'].iloc[i] = sig['mask'].iloc[i] + '_' + str(sig['MAF'].iloc[i])\n",
    "        sig['ID2'].iloc[i] = ID.iloc[i]\n",
    "\n",
    "        # Get regenie log url\n",
    "        phe = sig['pheno'].iloc[i]\n",
    "        if sig['CHROM'].iloc[i]!='X':\n",
    "            chro = str(int(sig['CHROM'].iloc[i]))\n",
    "        else:\n",
    "            chro = 'X'\n",
    "        ty = sig['pheno_type'].iloc[i]\n",
    "        log = \"/path/\"+ty+\"/\"+phe+\"/step2_\"+chro+\".log\"\n",
    "\n",
    "        # Read log info and get the number of cases and controls if binary\n",
    "        if ty =='binary':\n",
    "            with open(log, 'rb') as f:\n",
    "                contents = f.read().decode('utf-8')\n",
    "                num_cases = int(re.search(r'(\\d+) cases', contents).group(1))\n",
    "                num_controls = int(re.search(r'(\\d+) controls', contents).group(1))\n",
    "        else:\n",
    "            with open(log, 'rb') as f:\n",
    "                contents = f.read().decode('utf-8')\n",
    "                num_cases = int(re.search(r'(\\d+) observations', contents).group(1))\n",
    "                num_controls = ''\n",
    "\n",
    "        sig['cases'].iloc[i] = num_cases\n",
    "        sig['controls'].iloc[i] = num_controls\n",
    "        \n",
    "    return sig\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "interested-begin",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To read sig df if needed\n",
    "sig = r\"./path/all_sigs.tsv\"\n",
    "sig = pd.read_csv(sig, sep=\"\\t\")\n",
    "\n",
    "def conv(x):\n",
    "    try:\n",
    "        return int(float(x))\n",
    "    except ValueError:\n",
    "        return x\n",
    "    \n",
    "sig['CHROM'] = sig['CHROM'].apply(conv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ethical-coordination",
   "metadata": {},
   "outputs": [],
   "source": [
    "def near_gen(sig):\n",
    "    '''\n",
    "    Function for getting nearest protein-coding gene.\n",
    "    '''\n",
    "    sig['pos_st']=''\n",
    "    sig['pos_en']=''\n",
    "    sig['ensembl_gene_id']='N/A'\n",
    "    sig['gene_name']='N/A'\n",
    "\n",
    "    for i in range(len(sig)):\n",
    "        gene_f, sig['pos_st'].iloc[i], sig['pos_en'].iloc[i] = nearest_genes(sig['CHROM'].iloc[i], \n",
    "                                                                             sig['GENPOS'].iloc[i], \n",
    "                                                                             genes, 8e5)\n",
    "        if (gene_f is not None):\n",
    "            sig['ensembl_gene_id'].iloc[i] = gene_f['ensembl_gene_id']\n",
    "            sig['gene_name'].iloc[i] = gene_f['gene_name']\n",
    "            \n",
    "    return sig\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "entire-briefing",
   "metadata": {},
   "outputs": [],
   "source": [
    "def coor(sig):\n",
    "    '''\n",
    "    Function for getting coordinates  for manhattan plot.\n",
    "    '''\n",
    "    sig['pos'] = sig['pos'].astype(int)\n",
    "    sig = sig.sort_values(by=['pos'])\n",
    "    x_coor = []\n",
    "    cur_chrom=None\n",
    "    cur_pos=None\n",
    "    cur_x=0\n",
    "    for i, row in sig.iterrows():\n",
    "        if row['CHROM']!= cur_chrom:\n",
    "            cur_x+=1\n",
    "            cur_pos=row['pos']\n",
    "            cur_chrom=row['CHROM']\n",
    "        elif row['pos']-cur_pos>100000:\n",
    "            cur_x+=1\n",
    "            cur_pos=row['pos']\n",
    "        x_coor.append(cur_x)\n",
    "\n",
    "    sig['x_coor']= x_coor\n",
    "    return sig\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ordinary-tuesday",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read files for chrom and add info for binary phenos\n",
    "reg = None\n",
    "for chrom in range(1,23):\n",
    "    reg = None\n",
    "    reg = get_full_inf(pheno_b, reg, 'binary', int(chrom))\n",
    "    reg.to_csv(\"./path/binary\"+str(chrom)+\".tsv\", index=False, sep='\\t')\n",
    "    print(chrom)\n",
    "\n",
    "reg = None\n",
    "chrom='X'\n",
    "reg = get_full_inf(pheno_b, reg, 'binary', chrom)\n",
    "reg.to_csv(\"./path/binary\"+str(chrom)+\".tsv\", index=False, sep='\\t')\n",
    "print(chrom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "final-holly",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read files for chrom and add info for quantitative phenos\n",
    "reg = None\n",
    "chrom='X'\n",
    "reg = get_full_inf(pheno_q, reg, 'quan', chrom)\n",
    "reg.to_csv(\"./path/quan\"+str(chrom)+\".tsv\", index=False, sep='\\t')\n",
    "print(chrom)\n",
    "\n",
    "\n",
    "reg = None\n",
    "for chrom in range(1,23):\n",
    "    reg = None\n",
    "    reg = get_full_inf(pheno_q, reg, 'quan', int(chrom))\n",
    "    reg.to_csv(\"./path/quan\"+str(chrom)+\".tsv\", index=False, sep='\\t')\n",
    "    print(chrom)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "insured-benjamin",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in binary and quantitative info for chrom and merge and save\n",
    "\n",
    "for chrom in range(1,23):\n",
    "    reg = pd.read_csv(\"./path/binary\"+str(chrom)+\".tsv\", sep=\"\\t\")\n",
    "    print(len(reg))\n",
    "    reg2 = pd.read_csv(\"./path/quan\"+str(chrom)+\".tsv\", sep=\"\\t\")\n",
    "    print(len(reg2))\n",
    "    reg = reg.append(reg2).reset_index(drop=True).drop(['index'],axis=1)\n",
    "    print('app')\n",
    "    reg.to_csv(\"./path/\"+str(chrom)+\".tsv\", index=False, sep='\\t')\n",
    "    print(chrom)\n",
    "    reg=None\n",
    "    reg2=None\n",
    "    \n",
    "chrom='X'\n",
    "reg = pd.read_csv(\"./path/binary\"+str(chrom)+\".tsv\", sep=\"\\t\")\n",
    "print(len(reg))\n",
    "reg2 = pd.read_csv(\"./path/quan\"+str(chrom)+\".tsv\", sep=\"\\t\")\n",
    "print(len(reg2))\n",
    "reg = reg.append(reg2).reset_index(drop=True).drop(['index'],axis=1)\n",
    "print('app')\n",
    "reg.to_csv(\"./path/\"+str(chrom)+\".tsv\", index=False, sep='\\t')\n",
    "print(chrom)\n",
    "reg=None\n",
    "reg2=None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "conscious-inspector",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get coor for all chroms\n",
    "add = 0\n",
    "for chrom in range(1,23):\n",
    "    reg = None\n",
    "    rege = r\"./path/\"+str(chrom)+\".tsv\"\n",
    "    size=1000000\n",
    "\n",
    "    it = pd.read_csv(rege,chunksize=size, sep='\\t')\n",
    "    for chunk in it:\n",
    "        reg = pd.concat([reg, chunk])\n",
    "        print(len(reg))\n",
    "    \n",
    "    reg = coor(reg)\n",
    "    print('coor')\n",
    "    reg['x_coor']=reg['x_coor'] + add\n",
    "    add = reg['x_coor'].iloc[-1]\n",
    "    print('add')\n",
    "    reg.to_csv(\"./path/co\"+str(chrom)+\".tsv\", index=False, sep='\\t')\n",
    "    print(chrom) \n",
    "    print(add)\n",
    "    \n",
    "    \n",
    "chrom = 'X'\n",
    "reg = None\n",
    "rege = r\"./path/\"+str(chrom)+\".tsv\"\n",
    "size=1000000\n",
    "\n",
    "it = pd.read_csv(rege,chunksize=size, sep='\\t')\n",
    "for chunk in it:\n",
    "    reg = pd.concat([reg, chunk])\n",
    "    print(len(reg))\n",
    "        \n",
    "reg = coor(reg)\n",
    "print('coor')\n",
    "reg['x_coor']=reg['x_coor'] + add\n",
    "add = reg['x_coor'].iloc[-1]\n",
    "print('add')\n",
    "reg.to_csv(\"./path/co\"+str(chrom)+\".tsv\", index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "prospective-obligation",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make and save sig for each chrom\n",
    "for chrom in range(1,23):\n",
    "    reg = None\n",
    "    rege = r\"./path/co\"+str(chrom)+\".tsv\"\n",
    "    size=1000000\n",
    "\n",
    "    it = pd.read_csv(rege,chunksize=size, sep='\\t')\n",
    "    for chunk in it:\n",
    "        reg = pd.concat([reg, chunk[chunk['LOG10P']>5]])\n",
    "        print(len(reg))\n",
    "        \n",
    "    reg = near_gen(reg)\n",
    "    print('near gene')\n",
    "    reg = other_inf(reg)\n",
    "    print('other inf')\n",
    "    \n",
    "    reg.to_csv(\"./path/sig\"+str(chrom)+\".tsv\", index=False, sep='\\t')\n",
    "    print(chrom)\n",
    "    \n",
    "chrom = 'X'\n",
    "reg = None\n",
    "rege = r\"./path/co\"+str(chrom)+\".tsv\"\n",
    "size=1000000\n",
    "\n",
    "it = pd.read_csv(rege,chunksize=size, sep='\\t')\n",
    "for chunk in it:\n",
    "    reg = pd.concat([reg, chunk[chunk['LOG10P']>5]])\n",
    "    print(len(reg))\n",
    "        \n",
    "reg['CHROM'] = 'X'\n",
    "reg = near_gen(reg)\n",
    "print('near gene')\n",
    "reg = other_inf(reg)\n",
    "print('other inf')\n",
    "    \n",
    "reg.to_csv(\"./path/sig\"+str(chrom)+\".tsv\", index=False, sep='\\t')\n",
    "print(chrom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "official-muscle",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add all sigs to one files and save\n",
    "reg = None\n",
    "reg2 = None\n",
    "for chrom in range(1,23):\n",
    "    reg2 = pd.read_csv(\"./path/sig\"+str(chrom)+\".tsv\", sep=\"\\t\")\n",
    "    reg = pd.concat([reg, reg2])\n",
    "    print(chrom)\n",
    "    reg2=None\n",
    "\n",
    "chrom='X'\n",
    "reg2 = pd.read_csv(\"./path/sig\"+str(chrom)+\".tsv\", sep=\"\\t\")\n",
    "reg = pd.concat([reg, reg2])\n",
    "print(chrom)\n",
    "reg2=None    \n",
    "\n",
    "reg.to_csv(\"./path/all_sigs.tsv\", index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "understanding-dallas",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fill in snp info for sig and save final version\n",
    "sig['num_var'] = ''\n",
    "sig['snp_start_pos'] = ''\n",
    "sig['snp_end_pos'] = ''\n",
    "sig=snp_inf(pheno_q, sig, 'quan')\n",
    "sig=snp_inf(pheno_b, sig, 'binary')\n",
    "sig.to_csv(\"./path/full_sigs.tsv\", index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "awful-moisture",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read last file in if needed\n",
    "sig = r\"./path/full_sigs.tsv\"\n",
    "sig = pd.read_csv(sig, sep=\"\\t\")\n",
    "\n",
    "def conv(x):\n",
    "    try:\n",
    "        return int(float(x))\n",
    "    except ValueError:\n",
    "        return x\n",
    "    \n",
    "sig['CHROM'] = sig['CHROM'].apply(conv)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
